tests/test_musllinux.py

```diff
@@ -0,0 +1,123 @@
+        (MUSL_INVALID, None),
+        (MUSL_UNKNOWN, None),
+    ],
+    ids=["amd64-1.2.2", "i386-1.2.1", "aarch64-1.1.24", "invalid", "unknown"],
+)
+def test_parse_musl_version(output, version):
+    assert _parse_musl_version(output) == version
+
+
+@pytest.mark.parametrize(
+    "executable, location",
+    [
+        (BINARY_GLIBC_X86_64, None),
+        (BINARY_MUSL_X86_64, "/lib/ld-musl-x86_64.so.1"),
+        (BINARY_MUSL_I386, "/lib/ld-musl-i386.so.1"),
+        (BINARY_MUSL_AARCH64, "/lib/ld-musl-aarch64.so.1"),
+    ],
+    ids=["glibc", "x86_64", "i386", "aarch64"],
+)
+def test_parse_ld_musl_from_elf(executable, location):
+    with executable.open("rb") as f:
+        assert _parse_ld_musl_from_elf(f) == location
+
+
+@pytest.mark.parametrize(
+    "data",
+    [
+        # Too short for magic.
+        b"\0",
+        # Enough for magic, but not ELF.
+        b"#!/bin/bash" + b"\0" * 16,
+        # ELF, but unknown byte declaration.
+        b"\x7fELF\3" + b"\0" * 16,
+    ],
+    ids=["no-magic", "wrong-magic", "unknown-format"],
+)
+def test_parse_ld_musl_from_elf_invalid(data):
+    assert _parse_ld_musl_from_elf(io.BytesIO(data)) is None
+
+
+@pytest.mark.parametrize(
+    "head",
+    [
+        25,  # Enough for magic, but not the section definitions.
+        58,  # Enough for section definitions, but not the actual sections.
+    ],
+)
+def test_parse_ld_musl_from_elf_invalid_section(head):
+    data = BINARY_MUSL_X86_64.read_bytes()[:head]
+    assert _parse_ld_musl_from_elf(io.BytesIO(data)) is None
+
+
+def test_parse_ld_musl_from_elf_no_interpreter_section():
+    with BINARY_MUSL_X86_64.open("rb") as f:
+        data = f.read()
+
+    # Change all sections to *not* PT_INTERP.
+    unpacked = struct.unpack("16BHHIQQQIHHH", data[:58])
+    *_, e_phoff, _, _, _, e_phentsize, e_phnum = unpacked
+    for i in range(e_phnum + 1):
+        sb = e_phoff + e_phentsize * i
+        se = sb + 56
+        section = struct.unpack("IIQQQQQQ", data[sb:se])
+        data = data[:sb] + struct.pack("IIQQQQQQ", 0, *section[1:]) + data[se:]
+
+    assert _parse_ld_musl_from_elf(io.BytesIO(data)) is None
+
+
+@pytest.mark.parametrize(
+    "executable, output, version",
+    [
+        (MUSL_DIR.joinpath("does-not-exist"), "error", None),
+        (BINARY_GLIBC_X86_64, "error", None),
+        (BINARY_MUSL_X86_64, MUSL_AMD64_122, _MuslVersion(1, 2)),
+        (BINARY_MUSL_I386, MUSL_I386_121, _MuslVersion(1, 2)),
+        (BINARY_MUSL_AARCH64, MUSL_AARCH64_1124, _MuslVersion(1, 1)),
+    ],
+    ids=["does-not-exist", "glibc", "x86_64", "i386", "aarch64"],
+)
+def test_get_musl_version(monkeypatch, executable, output, version):
+    def mock_run(*args, **kwargs):
+        return collections.namedtuple("Proc", "stderr")(output)
+
+    run_recorder = pretend.call_recorder(mock_run)
+    monkeypatch.setattr(_musllinux.subprocess, "run", run_recorder)
+
+    assert _get_musl_version(str(executable)) == version
```

Comment:
I just realised. I used `pretend` because this test should check the subprocess call is invoked with the correct path, but forgot to actually write the assertions.